dbcrossbarlib/src/drivers/shopify/local_data.rs

```diff
@@ -0,0 +1,276 @@
+use serde::Deserialize;
+use serde_json::Value;
+use std::{collections::HashMap, str::FromStr};
+use tokio::{
+    sync::mpsc::Sender,
+    time::{delay_for, Duration},
+};
+
+use super::{json_to_csv::write_rows, ShopifyLocator};
+use crate::common::*;
+use crate::credentials::CredentialsManager;
+use crate::tokio_glue::{box_stream_once, bytes_channel, SendResultExt};
+
+pub(crate) async fn local_data_helper(
+    ctx: Context,
+    source: ShopifyLocator,
+    shared_args: SharedArguments<Unverified>,
+    source_args: SourceArguments<Unverified>,
+) -> Result<Option<BoxStream<CsvStream>>> {
+    let shared_args = shared_args.verify(ShopifyLocator::features())?;
+    let _source_args = source_args.verify(ShopifyLocator::features())?;
+    let schema = shared_args.schema().to_owned();
+
+    // Get our HTTPS URL, and specify that we always want the maximum number of records per request, because
+    // we're rate-limited per request, not per-record.
+    let mut url = source.to_https_url()?;
+    url.query_pairs_mut().append_pair("limit", "250").finish();
+
+    // Get our credentials.
+    let creds = CredentialsManager::singleton()
+        .get("shopify_secret")
+        .await?;
+    let auth_token = creds.get("auth_token")?.to_owned();
+
+    // Loop over pages until we run out.
+    let mut include_headers = true;
+    let worker_ctx = ctx.clone();
+    let (mut sender, receiver) = bytes_channel(1);
+    let worker: BoxFuture<()> = async move {
+        let client = Client::new();
+        let mut next_url = url.clone();
+        loop {
+            // Query Shopify and forward any errors to our consumer
+            let result = get_shopify_response(
+                &worker_ctx,
+                &client,
+                next_url,
+                auth_token.to_owned(),
+            )
+            .await;
+            let resp = match result {
+                Ok(resp) => resp,
+                Err(err) => {
+                    error!(
+                        worker_ctx.log(),
+                        "ERROR: {}",
+                        err.display_causes_without_backtrace(),
+                    );
+                    sender.send(Err(err)).await.map_send_err()?;
+                    return Ok(());
+                }
+            };
+
+            // Convert our data to CSV and send it, bailing if we hit an error.
+            if let Err(err) = convert_rows_to_csv_and_send(
+                &mut sender,
+                &schema,
+                resp.rows,
+                include_headers,
+            )
+            .await
+            {
+                sender.send(Err(err)).await.map_send_err()?;
+                return Ok(());
+            }
+            include_headers = false;
+
+            // Figure out what to do next.
+            if let Some(next_page_url) = resp.next_page_url {
+                next_url = next_page_url;
+
+                // If we're starting to overheat, wait a full second, giving enough
```

Comment:
we'll have to see how well this flows

Reply:
On average, we replenish 2 tokens/sec (or 4 tokens/second). There's also a burst pool. If we've used up less than half the burst pool, we run once a second and use as many tokens as we can until the burst pool is half full again.